### ISSUE TITLE
Problem with UMI Error Correction Using Bayesian Method

### ISSUE THREAD

**Issue by darneson** [hms-dbmi/dropEst#38]
Hello,

I am trying to do UMI error correction using the Bayesian correction method (I also want to note that the other correction methods used here: https://vpetukhov.github.io/dropEstAnalysis/umi_correction__umi_bmmc1.html work with no issues). When I run something similar to:

CorrectUmiSequenceErrors(method='Bayesian', return='reads',
collisions.info=collisions_info, umi.probabilities=umi_probs,
verbosity.level=2, mc.cores=30)

from the DropEstAnalysis site (https://vpetukhov.github.io/dropEstAnalysis/umi_correction__umi_bmmc1.html) I get the following error:

Error in `$<-.data.frame`(`*tmp*`, "IsMerged", value = logical(0)) : 
  replacement has 0 rows, data has 2

The actual code that I run (using my own DropSeq data which has UMI and cell BC quality tags) is:
corrected_reads$Bayesian <- holder$reads_per_umi_per_cell %>% 
	CorrectUmiSequenceErrors(method='Bayesian', return='reads', 
	collisions.info=collisions_info, umi.probabilities=umi_probs,
	verbosity.level=2, mc.cores=NULL)

Walking through the various functions, I am able to get to the **CorrectUmiSequenceErrorsBayesian** function, which is where the error occurs.

This step works ok:
clf <- try(**TrainNBClassifier**(reads.per.umi.per.cb, adj.umi.num=adj.umi.num, quality.quants.num=quality.quants.num,  mc.cores=mc.cores))

The error occurs here:
 filt.genes <- plapply(reads.per.umi.per.cb, **FilterUmisInGene**, clf, correction.info$dp.matrices,
                        correction.info$neighb.prob.index, collisions.info, mc.cores=mc.cores)

If I look inside the **FilterUmisInGene** function:
I see that the error occurs in the 10th entry in my reads.per.umi.per.cb object which looks as follows:

holder$reads_per_umi_per_cell$reads_per_umi[[10]]
$CCCCCGCT
$CCCCCGCT[[1]]
[1] 1

$CCCCCGCT[[2]]
[1] 41 41 41 41 41 41 41 41

$CCCCCGCG
$CCCCCGCG[[1]]
[1] 1

$CCCCCGCG[[2]]
[1] 41 41 41 41 41 41 41 41

Within **FilterUmisInGene**, the error occurs in:
predictions <- **PredictBayesian**(classifier, classifier.df, cur.gene[not.filtered.umis], dp.matrices,
                                   neighbours.prob.index, size.adj)

my **classifier.df** looks as follows (when I look here, I am guessing maybe this is where the problems arise, they are both one edit distance away from the other with the same quality and sam quality and error probs):

	Base   Target MinRpU MaxRpU Quality RealQualityProb ErrorQualityProb
1 CCCCCGCG CCCCCGCT      1      1      41      -0.6379865       -0.3912805
2 CCCCCGCT CCCCCGCG      1      1      41      -0.6379865       -0.3912805

**classifier.dfs** object is as follows:

$CCCCCGCG
  MinRpUCS MaxRpU   Target  ErrorProb   RealProb LargerNum
2        1      1 CCCCCGCG -0.3912805 -0.6379865         0

$CCCCCGCT
  MinRpUCS MaxRpU   Target  ErrorProb   RealProb LargerNum
1        1      1 CCCCCGCT -0.3912805 -0.6379865         0

**err.prior.probs** is as follows (where I assume the problems arise):

$CCCCCGCG
 1  2 
NA NA 

$CCCCCGCT
 1  2 
NA NA 

**real.prior.probs**  is as follows:

$CCCCCGCG
[1] 0.999478247 0.000521753

$CCCCCGCT
[1] 0.9997391235 0.0002608765

When we put this into, there is a problem in the **ErrorsNumMle** function:

classifier.df$IsMerged <- mapply(function(e.p, r.p, df)
    **ErrorsNumMle**(e.p, r.p, df$ErrorProb, df$RealProb, max.adj.num=classifier$MaxAdjacentUmisNum,
                 larger.num=df$LargerNum[1]) >= 1:nrow(df),
    err.prior.probs, real.prior.probs, classifier.dfs, SIMPLIFY=F) %>% unlist()

The problem is at the end of the function:

return(which.max(log(**prior.error.prob**) + log(rev(prior.real.prob)) + error.part.prob + real.part.prob) - 1)

Because the prior.error.prob is **NA NA** this makes everything else inside **NA NA** and end up returning **numeric(0)**

Tldr: Do you have a suggested way to deal with UMIs which are a single edit distance away and have the same quality scores (which I assume results in the **NA**)? Should I just return both (e.g. **return(cur.gene[not.filtered.umis])** )?

If I have missed the mark here and there is another underlying issue, I would greatly appreciate any help you can offer.

**Comment by VPetukhov** [comment]
Hi Douglas,

Thank you again for digging into the code. 
Indeed, function `ResolveUmiDependencies` is responsible for dealing with situation when two UMIs are merged into each other. Even if probability data frames are completely similar. And normally it shouldn't cause NA in probability values. I can suppose that it happens because of division by zero in line 28 of file umi_classifier.R, function `ErrorProbsGivenRlRs`. Could you please check if it's true? And if it is, can you please trace all variables within this function to understand, which one lead to zeros in `probs.subset`?

**Comment by VPetukhov** [comment]
Will reopen it if someone report more information for debug.

**Comment by VPetukhov** [comment]
Should be fixed in 6f7ac99
